ruby/private/providers.bzl

```diff
@@ -90,6 +90,29 @@ def get_transitive_runfiles(runfiles, srcs, data, deps):
             transitive_runfiles.append(target[DefaultInfo].default_runfiles)
     return runfiles.merge_all(transitive_runfiles)
 
+def get_transitive_input_targets(data, deps):
+    """Obtain the inputs for a target and its transitive dependencies.
+
+    Args:
+        data: a list of data targets
+        deps: a list of dependency targets
+    Returns:
+        a collection of the transitive input targets
+    """
+    input_targets = []
+    transitive_data = [dep[RubyFilesInfo].transitive_data for dep in deps]
+    transitive_deps = [dep[RubyFilesInfo].transitive_deps for dep in deps]
+    if len(transitive_data) > 0:
+        transitive_data = transitive_data[0].to_list()
```

Comment:
Hmm, two questions:
1. What's the significance of only taking the first element of `transitive_data`?
2. Calling `to_list()` is [generally discouraged](https://bazel.build/rules/performance#avoid-depset-to-list).

Do other rulesets typically call `to_list()` to allow interpolation in `env` or require a separate explicit set of targets to be passed through if the user wishes to use interpolation?

It looks like you're doing this so that you can check `type(x) == "Target"`. When isn't that the case for these depsets?

Reply:
Frankly, I'm not sure how this should work with depsets. The goal is to allow accumulating srcs/data from transitive dependencies to avoid having to add direct dependencies for each target:

```bazel
rb_library(
  name = "bar",
  srcs = ["baz"],
)

rb_binary(
  name = "foo",
  env = {"FOO": "$(location :baz)"},
  deps = ":bar",
}
```

Without this code, I'd have to add `baz` to `data` attribute of the `foo` target.  This is the only hacky way I've found to do so, but I would be glad to refactor into something more meaningful.

> What's the significance of only taking the first element of transitive_data?

It's always a list with a single element - a depset. Sometimes the list can be empty, so I'm having extra len > 0 check,

> Do other rulesets typically call to_list() to allow interpolation in env or require a separate explicit set of targets to be passed through if the user wishes to use interpolation?

@alexeagle Could you chime in and let us know if there is a good example in other rulesets we could follow?

> It looks like you're doing this so that you can check type(x) == "Target". When isn't that the case for these depsets?

Sometimes the input is a source file, leading to errors when passing to `ctx.expand_location`. I haven't debugged much what are cases there.

Reply:
Gotcha. I'll check this branch out and poke at it a bit...

Reply:
> The goal is to allow accumulating srcs/data from transitive dependencies to avoid having to add direct dependencies for each target:

OK, so, I think this is just something that isn't possible without sacrificing on performance by expanding these depsets 😅

Take for example `rules_js` which has a `js_run_binary` macro that dispatches to `@aspect_bazel_lib`'s `run_binary`:
https://github.com/aspect-build/rules_js/blob/58a5a5b0f7480845cb21086794da36d801ca0ffd/js/private/js_run_binary.bzl#L357-L370

The `run_binary` rule eventually boils down to calling this `expand_locations` helper which documents this limitation: https://github.com/aspect-build/bazel-lib/blob/197b2da974b7adc0718f77d2f7d532d8136667dc/lib/private/expand_locations.bzl#L6-L8.

That being said, I don't think it's a terrible limitation that if you want to expand something in your `env`, you need to directly depend on it.

Reply:
Ok, I will try reverting the transitive dependencies logic, though I still need to somehow filter out `File` objects from inputs.

Reply:
That comment is correct, to reference a label in an expansion you must also provide that label in an attribute on the same target with the expansion. It's not much hassle since it's local and symmetric - you don't have to list any unused transitives.

`to_list()` just need to take care to call only once at the terminal site that the list is to be iterated. There are some APIs (like ctx.actions.Args) that can directly accept a depset and delay the `to_list` until needed - but many APIs do require flattening yourself. Just take care not to do that in a loop, etc.